### ISSUE TITLE
goofys not connecting on Raspberry PI 3?

### ISSUE THREAD

**Issue by masterchiefaragorn** [kahing/goofys#387]
Hi,

I am trying to use goofys (on CentOS 7 on a Raspberry PI 3, in case it matters).

I am trying to mount "mybucket".  I've ensured my credentials file is correct and in sync with my IAM user on AWS.  I've asked AWS Support to debug this, and they are unsure why goofys is making a HEAD request to the "lfdh160d9ptwjwqq6wps8csriotsjjes" object:

    HEAD /mybucket/lfdh160d9ptwjwqq6wps8csriotsjjes HTTP/1.1

Here's the full output from goofys:

```
> -bash-4.2# goofys --debug_fuse --debug_s3 -f mybucket /root/backups/p001
> 2018/12/18 22:14:34.202125 s3.DEBUG HEAD https://s3.amazonaws.com/mybucket = 301 [us-west-2]
> 2018/12/18 22:14:34.202667 s3.INFO Switching from region 'us-east-1' to 'us-west-2'
> 2018/12/18 22:14:34.207409 s3.DEBUG DEBUG: Request s3/HeadObject Details:
> ---[ REQUEST POST-SIGN ]-----------------------------
> HEAD /mybucket/lfdh160d9ptwjwqq6wps8csriotsjjes HTTP/1.1
> Host: s3-us-west-2.amazonaws.com
> User-Agent: aws-sdk-go/1.8.25 (go1.9; linux; arm)
> Authorization: AWS4-HMAC-SHA256 Credential=MYKEYMYKEYMYKEYMYKEY/20181218/us-west-2/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=3ce6686cce9b317702af1107df13b854dcb9b44b8c6a307bf901dbf2f2692975
> X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
> X-Amz-Date: 20181218T221434Z
> 
> 
> -----------------------------------------------------
> 2018/12/18 22:14:34.721700 s3.DEBUG DEBUG: Response s3/HeadObject Details:
> ---[ RESPONSE ]--------------------------------------
> HTTP/1.1 404 Not Found
> Transfer-Encoding: chunked
> Content-Type: application/xml
> Date: Tue, 18 Dec 2018 22:14:33 GMT
> Server: AmazonS3
> X-Amz-Id-2: UBVFibRitmj70sses7J+KXr8mQk/O3bpLeD1FTS28lXp/lpuJF2f3ctdDbouO/nfV0mnN8zXx3g=
> X-Amz-Request-Id: 6A4EE1AE1F5EE022
> 
> 
> 
> -----------------------------------------------------
> 2018/12/18 22:14:34.722154 s3.DEBUG DEBUG: Validate Response s3/HeadObject failed, not retrying, error NotFound: Not Found
>         status code: 404, request id: 6A4EE1AE1F5EE022, host id: UBVFibRitmj70sses7J+KXr8mQk/O3bpLeD1FTS28lXp/lpuJF2f3ctdDbouO/nfV0mnN8zXx3g=
> 2018/12/18 22:14:34.727623 s3.DEBUG DEBUG: Request s3/ListMultipartUploads Details:
> ---[ REQUEST POST-SIGN ]-----------------------------
> GET /mybucket?uploads= HTTP/1.1
> Host: s3-us-west-2.amazonaws.com
> User-Agent: aws-sdk-go/1.8.25 (go1.9; linux; arm)
> Accept-Encoding: identity
> Authorization: AWS4-HMAC-SHA256 Credential=MYKEYMYKEYMYKEYMYKEY/20181218/us-west-2/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=68b611f20d80c5d6944c8be3920a3437e63f4bcaf9a68be9e38bc66e95565516
> X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
> X-Amz-Date: 20181218T221434Z
> 
> 
> -----------------------------------------------------
> 2018/12/18 22:14:34.836953 s3.DEBUG DEBUG: Response s3/ListMultipartUploads Details:
> ---[ RESPONSE ]--------------------------------------
> HTTP/1.1 200 OK
> Transfer-Encoding: chunked
> Content-Type: application/xml
> Date: Tue, 18 Dec 2018 22:14:35 GMT
> Server: AmazonS3
> X-Amz-Id-2: XeuQwk+RA1jHbNowDXwovv7WpFfg6TFLMXSo4bnHqN3k4Oybvv6gSftTPUgZP3YFFIIZojx+seM=
> X-Amz-Request-Id: 5E161BD45DD1E941
> 
> 
> -----------------------------------------------------
> 2018/12/18 22:14:34.837351 s3.DEBUG {
>   Bucket: "mybucket",
>   IsTruncated: false,
>   KeyMarker: "",
>   MaxUploads: 1000,
>   NextKeyMarker: "",
>   NextUploadIdMarker: "",
>   UploadIdMarker: ""
> }
> panic: unexpected OutMessage layout
> 
> goroutine 1 [running]:
> github.com/jacobsa/fuse/internal/buffer.(*OutMessage).Reset(0x112ae000)
>         /root/work/src/github.com/jacobsa/fuse/internal/buffer/out_message.go:67 +0x24
> github.com/jacobsa/fuse.(*Connection).getOutMessage(0x110a4930, 0x1128c000)
>         /root/work/src/github.com/jacobsa/fuse/freelists.go:60 +0x70
> github.com/jacobsa/fuse.(*Connection).ReadOp(0x110a4930, 0x110a4930, 0x76dbf3c0, 0x0, 0x1103ef00, 0x11062000, 0x76db8000)
>         /root/work/src/github.com/jacobsa/fuse/connection.go:385 +0x74
> github.com/jacobsa/fuse.(*Connection).Init(0x110a4930, 0x110a4930, 0x0)
>         /root/work/src/github.com/jacobsa/fuse/connection.go:121 +0x1c
> github.com/jacobsa/fuse.newConnection(0x6f3a50, 0x10f14098, 0x7eb1d813, 0x9, 0x0, 0x111546c0, 0x111546f0, 0x1, 0x0, 0x0, ...)
>         /root/work/src/github.com/jacobsa/fuse/connection.go:108 +0xc0
> github.com/jacobsa/fuse.Mount(0x7eb1d81d, 0x12, 0x6f0d80, 0x1103ed20, 0x1138fd84, 0x10f8b950, 0x10f96200, 0xa)
>         /root/work/src/github.com/jacobsa/fuse/mount.go:76 +0x290
> github.com/kahing/goofys/api.Mount(0x6f3a50, 0x10f14098, 0x7eb1d813, 0x9, 0x10f8b830, 0x0, 0x3aec8c, 0x1104fde8, 0x1104fdec)
>         /root/work/src/github.com/kahing/goofys/api/api.go:127 +0x3f8
> main.mount(0x6f3a50, 0x10f14098, 0x7eb1d813, 0x9, 0x10f8b7a0, 0x4800e3, 0x1, 0x10f7e988, 0x360ce0)
>         /root/work/src/github.com/kahing/goofys/main.go:117 +0x7c
> main.main.func1(0x10fd2160, 0x0, 0x0)
>         /root/work/src/github.com/kahing/goofys/main.go:213 +0x110
> github.com/urfave/cli.HandleAction(0x3f5c40, 0x10fbd5f0, 0x10fd2160, 0x0, 0x0)
>         /root/work/src/github.com/urfave/cli/app.go:501 +0xa8
> github.com/urfave/cli.(*App).Run(0x10f00fc0, 0x10f10080, 0x7, 0x8, 0x0, 0x0)
>         /root/work/src/github.com/urfave/cli/app.go:268 +0x448
> main.main()
>         /root/work/src/github.com/kahing/goofys/main.go:246 +0x134
> 
```

I can't tell if goofys is requesting an object it shouldn't be, or if it is doing the right thing, and if I need to work with AWS support more.

**Comment by masterchiefaragorn** [comment]
Ok.  After more research, I was able to get this working on an x86 machine.  So the issue is not goofys making a HEAD request to the "lfdh160d9ptwjwqq6wps8csriotsjjes" object.  The working goofys on the x86 machine does the same thing and it works.

I did a comparison of the two in debug mode, and the x86 one, which, works has these lines:

```
-----------------------------------------------------
2018/12/20 14:05:31.602672 fuse.DEBUG Op 0x00000001        connection.go:396] <- init
2018/12/20 14:05:31.602715 fuse.DEBUG Op 0x00000001        connection.go:479] -> OK ()
2018/12/20 14:05:31.602737 main.INFO File system has been successfully mounted.
2018/12/20 14:05:31.609175 s3.DEBUG DEBUG: Response s3/ListMultipartUploads Details:
---[ RESPONSE ]--------------------------------------
HTTP/1.1 200 OK

```

...whereas the one that fails only shows:

```
-----------------------------------------------------
2018/12/20 14:05:31.609175 s3.DEBUG DEBUG: Response s3/ListMultipartUploads Details:
---[ RESPONSE ]--------------------------------------
HTTP/1.1 200 OK

```

Does this appear to be something not quite right with fuse?  fuse was installed via yum, and is Version 2.9.2 (on the x86 machine that does work, it is CentOS 6 with fuse of 2.8.3).

**Comment by kahing** [comment]
I am not sure if the fuse library I use support non-x86.

**Comment by shilinki** [comment]
We actually just got this working for us yesterday, by recompiling the fuse library from jacobsa. As @masterchiefaragorn 's debug shows, the point where it would fail was within that project's code, at  

```
github.com/jacobsa/fuse/internal/buffer.(*OutMessage).Reset(0x112ae000)
>         /root/work/src/github.com/jacobsa/fuse/internal/buffer/out_message.go:67 +0x24
```

This points to this bit of code:

```
// Reset resets m so that it's ready to be used again. Afterward, the contents
// are solely a zeroed fusekernel.OutHeader struct.
func (m *OutMessage) Reset() {
	**// Ideally we'd like to write:
	//
	//     m.payloadOffset = 0
	//     m.header = fusekernel.OutHeader{}
	//
	// But Go 1.8 beta 2 generates bad code for this
	// (https://golang.org/issue/18370). Encourage it to generate the same code
	// as Go 1.7.4 did.
	if unsafe.Offsetof(m.payload) != 24 {
		panic("unexpected OutMessage layout")
	}**

	a := (*[3]uint64)(unsafe.Pointer(m))
	a[0] = 0
	a[1] = 0
	a[2] = 0
}
```
There was some kind of work-around behavior implemented there for Go 1.8 beta 2, which is what causes the failure. If the unsafe.Offset check is commented out and the m.payloadOffset and m.header lines uncommented back in, this works for us on ARMv7. I don't know that this is a perfect fix, but I hope it helps.

**Comment by masterchiefaragorn** [comment]
@kahing, @shilinki: Thanks for the replies.  I agree this issue lies with fuse.  Thanks for the help, @shilinki. 
Since posting this I have switched gears and gone with a Raspberry Pi 3+ which has worked, as it is aarch64.

If I even need to go back to ARMv7, I'll try out your suggestion.
